File: io/OutStream.java
package common.io;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

public strictfp interface OutStream {

	static OutStream getIns() {
		return new OutStreamDef();
	}

	static OutStream getAnimIns() {
		return new OutStreamAnim();
	}

	void accept(OutStream os);

	void flush(OutputStream fos) throws IOException;

	byte[] MD5();

	void terminate();

	InStream translate();

	void writeByte(byte n);

	void writeBytesB(byte[] s);

	void writeBytesI(byte[] s);

	void writeDouble(double n);

	void writeDoubles(double[] ints);

	void writeFloat(double n);

	void writeFloat(float n);

	void writeInt(int n);

	void writeIntB(int[] ints);

	void writeIntBB(int[][] ints);

	default void writeIntsN(int... ns) {
		for (int i : ns)
			writeInt(i);
	}

	void writeLong(long n);

	void writeShort(short n);

	void writeString(String str);

}

strictfp class OutStreamDef extends DataIO implements OutStream {

	private byte[] bs;
	private int index;

	public OutStreamDef() {
		bs = new byte[1024];
	}

	public OutStreamDef(int size) {
		bs = new byte[size];
		index = 0;
	}

	protected OutStreamDef(byte[] data) {
		bs = data;
		index = bs.length;
	}

	@Override
	public void accept(OutStream os) {
		if (!(os instanceof OutStreamDef))
			throw new BCUException("OutStream type mismatch");
		os.terminate();
		byte[] obs = ((OutStreamDef) os).bs;
		if (obs.length == 0)
			throw new BCUException("zero stream");
		writeInt(obs.length);
		check(obs.length);
		for (int i = 0; i < obs.length; i++)
			bs[index++] = obs[i];
	}

	public void concat(byte[] s) {
		check(s.length);
		for (byte b : s)
			fromByte(bs, index++, b);
	}

	@Override
	public void flush(OutputStream fos) throws IOException {
		terminate();
		fos.write(getSignature(bs.length));// signature
		fos.write(getBytes());
	}

	public byte[] getBytes() {
		return bs;
	}

	@Override
	public byte[] MD5() {
		try {
			MessageDigest mdi = MessageDigest.getInstance("MD5");
			mdi.update(getSignature(bs.length));
			return mdi.digest(bs);
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
		}
		return null;
	}

	public int pos() {
		return index;
	}

	@Override
	public void terminate() {
		if (index == bs.length)
			return;
		bs = Arrays.copyOf(bs, index);
	}

	@Override
	public String toString() {
		return "OutStreamDef " + size();
	}

	@Override
	public InStreamDef translate() {
		return new InStreamDef(translate(bs), 0, bs.length);
	}

	@Override
	public void writeByte(byte n) {
		check(1);
		fromByte(bs, index, n);
		index++;
	}

	@Override
	public void writeBytesB(byte[] s) {
		check(s.length + 1);
		writeByte((byte) s.length);
		for (byte b : s)
			writeByte(b);
	}

	@Override
	public void writeBytesI(byte[] s) {
		writeInt(s.length);
		check(s.length);
		for (byte b : s)
			writeByte(b);
	}

	@Override
	public void writeDouble(double n) {
		check(8);
		fromDouble(bs, index, n);
		index += 8;
	}

	@Override
	public void writeDoubles(double[] ints) {
		if (ints == null) {
			writeByte((byte) 0);
			return;
		}
		writeByte((byte) ints.length);
		for (double i : ints)
			writeDouble(i);
	}

	@Override
	public void writeFloat(double n) {
		writeFloat((float) n);
	}

	@Override
	public void writeFloat(float n) {
		check(4);
		fromFloat(bs, index, n);
		index += 4;
	}

	@Override
	public void writeInt(int n) {
		check(4);
		fromInt(bs, index, n);
		index += 4;
	}

	@Override
	public void writeIntB(int[] ints) {
		if (ints == null) {
			writeByte((byte) 0);
			return;
		}
		writeByte((byte) ints.length);
		for (int i : ints)
			writeInt(i);
	}

	@Override
	public void writeIntBB(int[][] ints) {
		if (ints == null) {
			writeByte((byte) 0);
			return;
		}
		writeByte((byte) ints.length);
		for (int[] i : ints)
			writeIntB(i);
	}

	@Override
	public void writeLong(long n) {
		check(8);
		fromLong(bs, index, n);
		index += 8;
	}

	@Override
	public void writeShort(short n) {
		check(2);
		fromShort(bs, index, n);
		index += 2;
	}

	@Override
	public void writeString(String str) {
		byte[] bts;
		bts = str.getBytes(StandardCharsets.UTF_8);
		writeBytesB(bts);
	}

	protected int size() {
		return bs.length;
	}

	private void check(int i) {
		if (index + i > bs.length * 2)
			bs = Arrays.copyOf(bs, index + i);
		else if (index + i > bs.length)
			bs = Arrays.copyOf(bs, bs.length * 2);
	}

}

strictfp class OutStreamAnim extends DataIO implements OutStream {

	private byte[] bs;
	private int index;

	public OutStreamAnim() {
		bs = new byte[1024];
	}

	@Override
	public void accept(OutStream os) {
		if (!(os instanceof OutStreamAnim))
			throw new BCUException("OutStream type mismatch");
		os.terminate();
		byte[] obs = ((OutStreamAnim) os).bs;
		if (obs.length == 0)
			throw new BCUException("zero stream");
		writeInt(obs.length);
		check(obs.length);
		for (int i = 0; i < obs.length; i++)
			bs[index++] = obs[i];
	}

	public void concat(byte[] s) {
		check(s.length);
		for (byte b : s)
			fromByte(bs, index++, b);
	}

	@Override
	public void flush(OutputStream fos) throws IOException {
		terminate();
		fos.write(getSignature(bs.length));// signature
		fos.write(getBytes());
	}

	public byte[] getBytes() {
		return bs;
	}

	@Override
	public byte[] MD5() {
		try {
			MessageDigest mdi = MessageDigest.getInstance("MD5");
			mdi.update(getSignature(bs.length));
			return mdi.digest(bs);
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
		}
		return null;
	}

	public int pos() {
		return index;
	}

	@Override
	public void terminate() {
		if (index == bs.length)
			return;
		bs = Arrays.copyOf(bs, index);
	}

	@Override
	public String toString() {
		return "OutStreamDef " + size();
	}

	@Override
	public InStreamAnim translate() {
		return new InStreamAnim(translate(bs), 0, bs.length);
	}

	@Override
	public void writeByte(byte n) {
		check(1);
		fromByte(bs, index, n);
		index++;
	}

	@Override
	public void writeBytesB(byte[] s) {
		check(s.length + 1);
		writeInt(s.length);
		for (byte b : s)
			writeByte(b);
	}

	@Override
	public void writeBytesI(byte[] s) {
		writeInt(s.length);
		check(s.length);
		for (byte b : s)
			writeByte(b);
	}

	@Override
	public void writeDouble(double n) {
		check(8);
		fromDouble(bs, index, n);
		index += 8;
	}

	@Override
	public void writeDoubles(double[] ints) {
		if (ints == null) {
			writeInt(0);
			return;
		}
		writeInt(ints.length);
		for (double i : ints)
			writeDouble(i);
	}

	@Override
	public void writeFloat(double n) {
		writeFloat((float) n);
	}

	@Override
	public void writeFloat(float n) {
		check(4);
		fromFloat(bs, index, n);
		index += 4;
	}

	@Override
	public void writeInt(int n) {
		check(4);
		fromInt(bs, index, n);
		index += 4;
	}

	@Override
	public void writeIntB(int[] ints) {
		if (ints == null) {
			writeInt(0);
			return;
		}
		writeInt(ints.length);
		for (int i : ints)
			writeInt(i);
	}

	@Override
	public void writeIntBB(int[][] ints) {
		if (ints == null) {
			writeInt( 0);
			return;
		}
		writeInt(ints.length);
		for (int[] i : ints)
			writeIntB(i);
	}

	@Override
	public void writeLong(long n) {
		check(8);
		fromLong(bs, index, n);
		index += 8;
	}

	@Override
	public void writeShort(short n) {
		check(2);
		fromShort(bs, index, n);
		index += 2;
	}

	@Override
	public void writeString(String str) {
		byte[] bts;
		bts = str.getBytes(StandardCharsets.UTF_8);
		writeBytesB(bts);
	}

	protected int size() {
		return bs.length;
	}

	private void check(int i) {
		if (index + i > bs.length * 2)
			bs = Arrays.copyOf(bs, index + i);
		else if (index + i > bs.length)
			bs = Arrays.copyOf(bs, bs.length * 2);
	}

}

File: io/OutStream.kt
package common.io

import common.io.DataIO
import java.io.IOException
import java.io.OutputStream
import java.io.UnsupportedEncodingException
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.*

@Strictfp
interface OutStream {
    fun accept(os: OutStream)

    @Throws(IOException::class)
    fun flush(fos: OutputStream)
    fun MD5(): ByteArray?
    fun terminate()
    fun translate(): InStream
    fun writeByte(n: Byte)
    fun writeBytesB(s: ByteArray)
    fun writeBytesI(s: ByteArray)
    fun writeDouble(n: Double)
    fun writeDoubles(ints: DoubleArray?)
    fun writeFloat(n: Double)
    fun writeFloat(n: Float)
    fun writeInt(n: Int)
    fun writeIntB(ints: IntArray?)
    fun writeIntBB(ints: Array<IntArray?>?)
    fun writeIntsN(vararg ns: Int) {
        for (i in ns) writeInt(i)
    }

    fun writeLong(n: Long)
    fun writeShort(n: Short)
    fun writeString(str: String)

    companion object {
        fun getIns(): OutStream? {
            return OutStreamDef()
        }
    }
}

@Strictfp
internal class OutStreamDef : DataIO, OutStream {
    private var bs: ByteArray
    private var index = 0

    constructor() {
        bs = ByteArray(1024)
    }

    constructor(size: Int) {
        bs = ByteArray(size)
        index = 0
    }

    constructor(data: ByteArray) {
        bs = data
        index = bs.size
    }

    override fun accept(os: OutStream) {
        if (os !is OutStreamDef) throw BCUException("OutStream type mismatch")
        os.terminate()
        val obs = os.bs
        if (obs.size == 0) throw BCUException("zero stream")
        writeInt(obs.size)
        check(obs.size)
        for (i in obs.indices) bs[index++] = obs[i]
    }

    fun concat(s: ByteArray) {
        check(s.size)
        for (b in s) DataIO.Companion.fromByte(bs, index++, b)
    }

    @Throws(IOException::class)
    override fun flush(fos: OutputStream) {
        terminate()
        fos.write(DataIO.Companion.getSignature(bs.size)) // signature
        fos.write(getBytes())
    }

    fun getBytes(): ByteArray {
        return bs
    }

    override fun MD5(): ByteArray? {
        try {
            val mdi: MessageDigest = MessageDigest.getInstance("MD5")
            mdi.update(DataIO.Companion.getSignature(bs.size))
            return mdi.digest(bs)
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
        }
        return null
    }

    fun pos(): Int {
        return index
    }

    override fun terminate() {
        if (index == bs.size) return
        bs = Arrays.copyOf(bs, index)
    }

    override fun toString(): String {
        return "OutStreamDef " + size()
    }

    override fun translate(): InStreamDef {
        return InStreamDef(DataIO.Companion.translate(bs), 0, bs.size)
    }

    override fun writeByte(n: Byte) {
        check(1)
        DataIO.Companion.fromByte(bs, index, n)
        index++
    }

    override fun writeBytesB(s: ByteArray) {
        check(s.size + 1)
        writeByte(s.size.toByte())
        for (b in s) writeByte(b)
    }

    override fun writeBytesI(s: ByteArray) {
        writeInt(s.size)
        check(s.size)
        for (b in s) writeByte(b)
    }

    override fun writeDouble(n: Double) {
        check(8)
        DataIO.Companion.fromDouble(bs, index, n)
        index += 8
    }

    override fun writeDoubles(ints: DoubleArray?) {
        if (ints == null) {
            writeByte(0.toByte())
            return
        }
        writeByte(ints.size.toByte())
        for (i in ints) writeDouble(i)
    }

    override fun writeFloat(n: Double) {
        writeFloat(n.toFloat())
    }

    override fun writeFloat(n: Float) {
        check(4)
        DataIO.Companion.fromFloat(bs, index, n)
        index += 4
    }

    override fun writeInt(n: Int) {
        check(4)
        DataIO.Companion.fromInt(bs, index, n)
        index += 4
    }

    override fun writeIntB(ints: IntArray?) {
        if (ints == null) {
            writeByte(0.toByte())
            return
        }
        writeByte(ints.size.toByte())
        for (i in ints) writeInt(i)
    }

    override fun writeIntBB(ints: Array<IntArray?>?) {
        if (ints == null) {
            writeByte(0.toByte())
            return
        }
        writeByte(ints.size.toByte())
        for (i in ints) writeIntB(i)
    }

    override fun writeLong(n: Long) {
        check(8)
        DataIO.Companion.fromLong(bs, index, n)
        index += 8
    }

    override fun writeShort(n: Short) {
        check(2)
        DataIO.Companion.fromShort(bs, index, n)
        index += 2
    }

    override fun writeString(str: String) {
        var bts: ByteArray
        try {
            bts = str.toByteArray(charset("UTF-8"))
        } catch (e: UnsupportedEncodingException) {
            bts = str.toByteArray()
            e.printStackTrace()
        }
        writeBytesB(bts)
    }

    protected fun size(): Int {
        return bs.size
    }

    private fun check(i: Int) {
        if (index + i > bs.size * 2) bs = Arrays.copyOf(bs, index + i) else if (index + i > bs.size) bs = Arrays.copyOf(bs, bs.size * 2)
    }
}
